text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,550 @@
+| Source Maps (.js.map)          | 15.87%                         |
+| Javascript code (.js)          | 6.63%                          |
+| Type Declaration files (.d.ts) | 6.62%                          |
+| README, etc (.md)              | 0.55%                          |
+| bundled npm dependencies       | 0.50%                          |
+| .ts-fixture                    | 0.27%                          |
+| .jsiirc                        | 0.26%                          |
+
+### Why should we _not_ do this?
+
+The main reason to not do this is that it will inevitably break up the
+monolothic `aws-cdk-lib` package in some manner, and we will lose some of the
+wonderful simplicity of releasing and using a single package. But, we can't keep
+everything in one place, and reduce the size enough to have a significant
+positive impact.
+
+### What is the technical solution (design) of this feature?
+
+There are different categories of files contributing to the size of aws-cdk-lib,
+and each category will have it’s own solution. The following sections cover the
+top five categories in size contribution in order from largest to smallest
+potential size.
+
+#### Lambda Layer Zip Files
+
+Use separate npm packages for each `lambda-layer-X` module, which the CDK CLI
+will automatically install if not already present in the `node_modules`
+directory. With this solution, customers will not have to figure out when they
+do or do not need to include these dependencies, and they will not be required
+to access a public endpoint besides npm during runtime of their CDK apps.
+Practically, this solution breaks down into the following steps.
+
+1. Publish v2 compatible versions of `@aws-cdk/lambda-layer-aws-cli`,
+   `@aws-cdk/lambda-layer-aws-cli`, `@aws-cdk/lambda-layer-node-proxy-agent` as
+   their own npm packages, separate from `aws-cdk-lib`. This publishing process
+   will be an extension of the existing AWS CDK V2 release machinery. Any
+   submodules with the prefix `lambda-layer` will be published the same way.
+2. Modify the `lambda-layer-X` submodules in `aws-cdk-lib` to not bundle these
+   large dependencies themselves, but instead reference the appropriate `Layer`
+   class from the packages in step 1. Since `aws-cdk-lib`'s source code is
+   auto-generated from the v1 submodules, this could be an autogenerated
+   wrapper. We will need to do something clever here to make sure customers do
+   not have IDE or compilation errors before the CDK CLI has an opportunity to
+   download and install the packages.
+3. Modify the CDK CLI to verify that the correct packages are available in the
+   `node_modules` directory during synthesis. If they are not available,
+   download them from npm.
+4. We need to keep the dependencies up-to-date to get security updates and new
+   features. Each Lambda Layer will need to be treated slightly differently.
+    1. `@aws-cdk/lambda-layer-kubectl` - This one has two dependencies to keep
```

Comment:
Why don't we have `@aws-cdk/lambda-layer-kubectl-120` and `@aws-cdk/lambda-layer-kubectl-122` etc?

Reply:
I didn't think about this. I was thinking it makes logical sense to group all versions of one dependency into the same package. But it might actually be simpler to manage if they are separate. What do you think will be the major difference between the two options?

Reply:
I've given this some more thought, and I think it will be best to have separate libraries for each minor version like you suggested. This way, the runtime only needs to download the kubectl version(s) that will actually be used. Updated the doc to reflect this.